Refactor async_turn_on() for ZHA Light.

Adminiuga opened the pull request:
## Description:
Use "move_to_level_with_on_off" if brightness or transition attributes
are present in the service call data, otherwise issue "On" Zigbee
command.
Allow brightness of 0 for service call -- effectively turning the light
off.
Send color commands only after the light was turned on.


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>
#21107 #17943 

## Checklist:
  - [X] The code change is tested and works locally.
  - [] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)


[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

dmulcahey reviewed (commented):
Just fix the test and I think this is good.

perosb commented:
Nice, I was just about to report an issue that when calling a `scene` with both color/temp and brightness only issues one of the commands (IKEA bulbs)
Hoping that this PR will fix it. I will test it later on at home.

Thanks! 👍

Adminiuga committed:
Refactor async_turn_on() for ZHA Light.

Use "move_to_level_with_on_off" if brightness or transition attributes
are present in the service call data, otherwise issue "On" Zigbee
command.
Allow brightness of 0 for service call -- effectively turning the light
off.
Send color commands only after the light was turned on.

perosb commented on homeassistant/components/zha/light.py:
```diff
@@ -146,8 +151,31 @@ def async_set_state(self, state):
+                level = min(254, brightness)
+            else:
+                level = self._brightness or 254
+            success = await self._level_listener.move_to_level_with_on_off(
```
Should this be `self._level_channel`?

Adminiuga commented:
I pushed new version. Still need to fix test. Hopefully tonight.

On Tue, 26 Feb 2019 at 01:51, Per Osbäck <notifications@github.com> wrote:

> *@perosb* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/zha/light.py
> <https://github.com/home-assistant/home-assistant/pull/21156#discussion_r260146081>
> :
>
> > @@ -146,8 +151,31 @@ def async_set_state(self, state):
>
>      async def async_turn_on(self, **kwargs):
>          """Turn the entity on."""
> -        duration = kwargs.get(light.ATTR_TRANSITION, DEFAULT_DURATION)
> -        duration = duration * 10  # tenths of s
> +        transition = kwargs.get(light.ATTR_TRANSITION)
> +        duration = transition * 10 if transition else DEFAULT_DURATION
> +        brightness = kwargs.get(light.ATTR_BRIGHTNESS)
> +
> +        if (brightness is not None or transition) and \
> +                self._supported_features & light.SUPPORT_BRIGHTNESS:
> +            if brightness is not None:
> +                level = min(254, brightness)
> +            else:
> +                level = self._brightness or 254
> +            success = await self._level_listener.move_to_level_with_on_off(
>
> Should this be self._level_channel?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/21156#pullrequestreview-207784857>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFjmcGozB4w4aV2NWam5ly0mwHKe1Z1iks5vRNmOgaJpZM4a_1bv>
> .
>

Adminiuga committed:
Fix zha.light tests.
